### PR TITLE
new svd2regs tool

### PR DESCRIPTION
**Pull request by stefanhoelzl**
### Pull Request Overview

This pull request adds a new tool to generate TockOS-conform Peripheral structs and bitfield_macros from SVD-files.

### Testing Strategy

This pull request was tested with the Kinetis MK64F12 SVD and the SIM, PMC peripheral.

### TODO or Help Wanted

- Use it with different MCUs/Peripherals and report bugs

### Documentation Updated

- [X] Kernel: Updated the relevant files in `/docs`, or no updates are required.
- [X] Userland: Added/updated the application README, if needed.

### Formatting

- [X] Ran `make formatall`.

[example_output.txt](https://github.com/helena-project/tock/files/1910580/example_output.txt)

### CODE REVIEW

**Comment by bradjc** [comment]
This is cool! I added a couple commits from when i was testing with `python svd2regs.py --mcu Nordic nrf52 TEMP`.

**Comment by stefanhoelzl** [comment]
Thanks!
Python 3 support is a good one

**Comment by stefanhoelzl** [comment]
Did also some tests with the GPIO and PORT peripheral (Kinetis K64F12).
In the PORT peripheral was some manual rework necessary, but this is because the SVD didn't mark some registers as grouped, so there is no way to determine, that these definitions can be the same.
And the enumeration names are way too long, I had to use the descriptions because the names given by the SVD are the same than the values (e.g. 0xFF,  1)
If someone finds some SVD during testing, maybe I can change it to use the names instead of descriptions.

File: tools/svd2regs.py [hunk 1]
```diff
@@ -0,0 +1,312 @@
...
+        return {
+            "name": peripheral.name,
+            "title": base.title(),
+            "base": peripheral.base_address,
+        }
+
+
+class PeripheralStruct(CodeBlock):
+    TEMPLATE = """{comment}
+#[repr(C, packed)]
```
Comment (niklasad1) [inline tools/svd2regs.py]: You shouldn't use `packed` here

File: tools/svd2regs.py [hunk 2]
```diff
@@ -0,0 +1,312 @@
...
+
+    @staticmethod
+    def fields(*args):
+        return {}
+
+
+class Includes(CodeBlock):
+    TEMPLATE = """
+use kernel::StaticRef;
+use kernel::common::regs::{{self, ReadOnly, ReadWrite, WriteOnly}};
```
Comment (niklasad1) [inline tools/svd2regs.py]: No need for double parentheses?

File: tools/svd2regs.py (hunk 1 above)
Comment (stefanhoelzl) [reply tools/svd2regs.py]: ok, why not?
Took how to build the definition from https://www.tockos.org/blog/2018/talking-tock-34/#register-and-bitfield-macros

File: tools/svd2regs.py (hunk 2 above)
Comment (stefanhoelzl) [reply tools/svd2regs.py]: it is a template string which gets formatted.
Places where values are inserted are marked with {}, so double parentheses do escape these to and get to single parentheses after inserting of the values.

When you look at other classes which derive from CodeBlock there are values to replace in the TEMPLATE.

Comment (niklasad1) [reply tools/svd2regs.py]: Ok, sorry!

File: tools/svd2regs.py [hunk 3]
```diff
@@ -0,0 +1,312 @@
...
+    """
+
+
+class PeripheralBaseDeclaration(CodeBlock):
+    TEMPLATE = """
+const {name}_BASE: StaticRef<{title}Registers> =
+    unsafe {{ StaticRef::new(0x{base:8X} as *const {title}Registers) }};
```
Comment (niklasad1) [inline tools/svd2regs.py]: I think unsafe is not needed here

File: tools/svd2regs.py [hunk 4]
```diff
@@ -0,0 +1,312 @@
...
+#
+#   With stdin pipe
+#     cat mcu.svd | svd2regs.py SIM --svd --fmt '--force' | tee src/mcu.rs
+#
+# Required Python Packages:
+#   cmsis-svd
+#   pydentifier
+#
+# Author: Stefan Hoelzl <stefan.hoelzl@posteo.de>
+
```
Comment (niklasad1) [inline tools/svd2regs.py]: add `#!/usr/bin/python3` please :)

**Comment by niklasad1** [comment]
I ran: `./svd2regs.py --mcu Nordic nrf52 RADIO`

File: tools/svd2regs.py (hunk 3 above)
Comment (niklasad1) [reply tools/svd2regs.py]: Yepp, you are right `StaticRef::new` is unsafe and therefore it should be unsafe. It is wrapper around raw pointer manipulations! I have not used it personally but sorry again!

**Comment by stefanhoelzl** [comment]
Ok, tested also with ```svd2regs.py --mcu Nordic nrf52 RADIO``` with the latest changes it works now.
Seems there are different ways in the SVD Format to define register sizes

**Comment by bradjc** [comment]
I can't get the `--svd` option to work. I keep getting:

```
Traceback (most recent call last):
  File "svd2regs.py", line 328, in <module>
    main()
  File "svd2regs.py", line 321, in main
    code = generate(*parse(args.peripheral, args.mcu, args.svd, args.group))
  File "svd2regs.py", line 260, in parse
    svd_parser = get_parser(mcu, svd)
  File "svd2regs.py", line 253, in get_parser
    return SVDParser(ET.fromstring(svd.read()))
  File "/usr/local/lib/python3.6/site-packages/cmsis_svd-0.5.dev0-py3.6.egg/cmsis_svd/parser.py", line 111, in __init__
    self._root = self._tree.getroot()
AttributeError: 'xml.etree.ElementTree.Element' object has no attribute 'getroot'
```

I had to change the `--svd` option to just a be a string and then read that with `ET.parse()`.

**Comment by bradjc** [comment]
Cool works now.

**Comment by bradjc** [comment]
Last commits were 5 days ago, probably time for last call on this one if anyone has any feedback.

Marked significant to encourage people to look at this new tool, but it isn't part of the kernel and won't break anything.

**Pull request merged**